File: 0078-subsets/0078-subsets.cpp
void dfs(vector<int>& v, vector<vector<int>> &ans,vector<int>& nums,int i) {
    if(i >= nums.size()) {
        ans.push_back(v);
        return;
    }
    v.push_back(nums[i]);
    dfs(v,ans,nums,i+1);
    v.pop_back();
    dfs(v,ans,nums,i+1);
}

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> v;
        dfs(v,ans,nums,0);
        return ans;
    }
};

File: 0078-subsets/0078-subsets.java
class Solution {
    public static void backtrack(int id,List<Integer> list,List<List<Integer>> ans,int[] nums) {
        if(id == nums.length) {
            ans.add(new ArrayList(list));
            return;
        }
        // take
        list.add(nums[id]);
        backtrack(id+1,list,ans,nums);
        // undo
        list.remove(list.size() - 1);
        // skip
        backtrack(id+1,list,ans,nums);
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> list = new ArrayList<>();

        backtrack(0,list,ans,nums);

        return ans;
    }
}